src/hotspot/share/opto/node.cpp

```diff
@@ -1588,25 +1588,53 @@ jfloat Node::getf() const {
 #ifndef PRODUCT
 
 // Call this from debugger:
+Node* old_root() {
+  Matcher* matcher = Compile::current()->matcher();
+  if (matcher != nullptr) {
+    Node* new_root = Compile::current()->root();
+    Node* old_root = matcher->find_old_node(new_root);
+    if (old_root != nullptr) {
+      return old_root;
+    }
+  }
+  tty->print("old_root: not found.\n");
+  return nullptr;
+}
+
+// Call this from debugger, search in same graph as n:
 Node* find_node(Node* n, const int idx) {
   return n->find(idx);
 }
 
-// Call this from debugger with root node as default:
+// Call this from debugger, search in new nodes:
 Node* find_node(const int idx) {
   return Compile::current()->root()->find(idx);
 }
 
-// Call this from debugger:
+// Call this from debugger, search in old nodes:
+Node* find_old_node(const int idx) {
+  Node* root = old_root();
+  assert(root != nullptr, "must have old_root() to find old nodes");
```

Comment:
@chhagedorn I added this `nullptr` because @vnkozlov asked for one. https://github.com/openjdk/jdk/pull/8567#discussion_r866966586
Is there an alternative to an assert?
I realize, that in `rr` the debuggin session is not stopped for me - it just unwinds - and I can continue debugging as if nothing happened. However in `gdb` this does crash the debugging session - at least by default. You can change that behavior with `set unwindonsignal on`.

There are other asserts that can be triggered. For example if we call `find_node(1)` from a non-compiler thread, we get an assert when internally `Compile::current()` comes across this

https://github.com/openjdk/jdk/blob/b849efdf154552903faaddd69cac1fe5f1ddf18a/src/hotspot/share/compiler/compilerThread.hpp#L63-L64

Reply:
I did not know about `set unwindonsignal on`, that's good trick! I agree that you can still trigger assertions while debugging but I think we should try to reduce that chance as much as we can with sane input values.